### PR TITLE
TBRANDS-106 & TBRANDS-126 : Product Gallery Desktop View

### PR DESCRIPTION
**Pull request by ecarlisle**
## Description
When the browser viewport is desktop-sized, clicking on images within the Product Gallery will produce a "Focused View." This view is essentially a modal overlay, allowing the user to browse larger product images. 
* [Figma Prototype](https://www.figma.com/proto/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=8294%3A130007&scaling=scale-down&page-id=8294%3A119303&starting-point-node-id=8294%3A119408&show-proto-sidebar=1)
* [Figma Design](https://www.figma.com/file/vWVGgZfZVBdSX7ACo3kmjr/Enhanced-Styling-Design-System?node-id=8020%3A112533)

## Jira Ticket

- Original Ticket : [TBRANDS-106](https://arcpublishing.atlassian.net/browse/TBRANDS-106)
- Follow-up Ticket: [TBRANDS-126](https://arcpublishing.atlassian.net/browse/TBRANDS-126)

### From TBRANDS-106
1. When a consumer user clicks on any image rendered as part of the Product Gallery, the Focused View deploys.
2. The Focused View fills the user's browser’s viewport. (Not Fullscreen API)
3. A user can scroll vertically through all gallery images.
4. A user can navigate to a specific image by clicking on a thumbnail in the left rail. 
5. A user sees a thumbnail highlighted that corresponds to the image they are viewing.
6. A user can close the Focused View by clicking on the close icon found at the top right.

#### BEHAVIOR (Responsive)
On desktop, clicking on a gallery media item will open the Gallery Focus for a more detailed view.

#### IMAGE
1. Aspect ratio should be square (1:1), at least 2000px x 2000px
2. Image should fill 100% width of the dedicated content area above, height extending beyond the viewport height
3. Scrolling over focused content area will reveal additional media

#### SCROLL (over focused media) 
* All media in the Gallery Focus View can be accessed via vertical scroll. 
* Scrolling over the focused media will reveal the next media item in the carousel. 
* Once that next item takes up the majority screen height, the corresponding thumbnail will be selected.
* If the thumbnail is not present in current carousel view, the carousel will shift to ensure the selected thumbnail is in view.
#### Thumbnails
1. Aspect ratio should be square (1:1), 100x100
2. Height of Thumbnail Bar should not exceed viewport height - 80px. Number of thumbnails displayed at a time is dependent on the viewport height. An additional thumbnail can be shown as long as the gap between elements and padding above and below is maintained.
3. Thumbnail bar is fixed so should maintain it’s position when focused area is vertically scrolled.

### Assumed Out of Scope
1. Video integration
2. Image zoom control
3. Product color selection

## Test Steps

1. `git checkout TBRANDS-106-product-gallery-focus-view` on local `arc-themes-blocks` repository.
2. `git checkout TBRANDS-106-product-gallery-focus-view` on local `arc-themes-feature-pack` repository.
3. In local `arc-themes-feature-pack` repository, enable the following lines to their local repositories:
    a. `THEME_COMPONENTS_REPO` (needed for change in `Image` component).
    b. `THEMES_BLOCKS_REPO`
4. Start feature pack locally with `npx fusion start -f -l @wpmedia/product-gallery-block`.
5. Create a new page in PageBuilder Editor:
    a. Set global content of page to `commerce-product` using `bounce-running-shoe` as the SKU.
    b. Add "Product Gallery - Arc Block" to page.
    c. Set "Display the product's featured image" to true.
6. Publish the page and view locally for best debugging.

#### Product Gallery at fullscreen (without Focused View overlay).
image

#### Focused View Overlay (Number of thumbnails statically set to 3)
image

#### Thumbnail Bar States (3)
image

## Dependencies or Side Effects

Feature Pack Styles - https://github.com/WPMedia/arc-themes-feature-pack/pull/359

## Author Checklist

_The author of the PR should fill out the following sections to ensure this PR is ready for review._

- [ ] Confirmed all the test steps a reviewer will follow above are working.
- [ ] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [ ] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [ ] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [ ] Confirmed this PR has unit test files
  - [ ] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
        please explain why (so that we can fix it whenever it gets refactored).
- [ ] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist

_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr.
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

### CODE REVIEW

File: .storybook/themes/commerce.scss [hunk 1]
```diff
@@ -803,6 +803,75 @@
...
+							width: 100%,
+						),
+						image: (
+							max-height: 100px,
+							max-width: 100px,
+						),
+						paragraph: (
+							font-family: var(--font-family-primary),
+							font-size: var(--heading-level-6-font-size),
+							font-weight: var(--global-font-size-4),
```
Comment (nschubach) [inline .storybook/themes/commerce.scss]: Font weights should use font-weight values.

File: blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+				</div>
+				<Stack
+					alignment="center"
+					direction="vertical"
+					className={`${BLOCK_CLASS_NAME}__focus-view-main-images`}
+				>
+					{productImages.map((image, index) => (
+						<div
+							style={{ flexBasis: "100vh", flexShrink: 0 }}
+							key={image._id}
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx]: `key`s should be unique:
image

Should just be a matter of appending the `_id`: `focus_main_view_${image._id}`

File: blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+							)
+							.map((image) => (
+								<Image
+									// used as part of a page design so empty string alt text
+									alt=""
+									className={`${BLOCK_CLASS_NAME}__focus-view-thumbnail-image${
+										image._id === productImages[selectedImageIndex]._id ? "--selected" : ""
+									}`}
+									height={100}
+									key={image._id}
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx]: `key`s should be unique.

Should just be a matter of appending the `_id`: `focus_view_thumbnail_${image._id}`

File: blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+					</Paragraph>
+					{shouldShowUpButton() ? (
+						<Button
+							accessibilityLabel="Previous Image"
+							onClick={() => setThumbnailBarStartIndex(thumbnailBarStartIndex - 1)}
+						>
+							<Icon name="ChevronUp" />
+						</Button>
+					) : (
+						<div className={`${BLOCK_CLASS_NAME}__focus-view-thumbnail-bar-spacer`} />
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx]: Instead of a spacer, why don't we make this container(`focus-view-thumbnail-bar`) a grid and set the row height?  The "padding" could be row 1, selected text could be row 2, button in row 3, stack in row 4, button in row 5, and the bottom padding is row 6.

```
product-gallery-focus-view-thumbnail-bar: (
    display: grid,
    grid-template-rows: var(--global-spacing-7) auto 1fr auto 1fr var(--global-spacing-7),
    grid-template-columns: var(--global-spacing-6) auto var(--global-spacing-6),
)
```

File: .storybook/themes/commerce.scss [hunk 5]
```diff
@@ -803,6 +803,75 @@
...
+					right: 0,
+					top: 0,
+				),
+				product-gallery-focus-view-thumbnail-bar: (
+					gap: var(--global-spacing-5),
+					padding-top: var(--global-spacing-7),
+					padding-right: var(--global-spacing-6),
+					padding-bottom: var(--global-spacing-7),
+					padding-left: var(--global-spacing-6),
+					text-align: center,
```
Comment (nschubach) [inline .storybook/themes/commerce.scss]: Technically this (`text-align`) should be on the paragraph.  If you want to center align the elements, with flex/grid, we should be using `align-items`/`justify-content` notation.

File: .storybook/themes/commerce.scss [hunk 6]
```diff
@@ -803,6 +803,75 @@
...
+					border-color: var(--global-white),
+					border-style: var(--global-border-style-1),
+					border-width: 0.5px,
+					opacity: 0.75,
+				),
+				product-gallery-focus-view-thumbnail-image-selected: (
+					border-color: var(--global-black),
+					border-style: var(--global-border-style-1),
+					border-width: 0.5px,
+				),
```
Comment (nschubach) [inline .storybook/themes/commerce.scss:874]: If selected is a modified state, you should only have to set the `border-color`

File: blocks/product-gallery-block/features/product-gallery/default.jsx [hunk 7]
```diff
@@ -43,63 +47,75 @@ export function ProductGalleryDisplay({
...
-												mediaCondition: "(min-width: 48rem)",
-											},
-									  ]),
-							]}
-						/>
-					</Carousel.Item>
-				);
-			})}
-		</Carousel>
+		<div>
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/default.jsx:50]: Why was this added?

Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/default.jsx]: This can be a fragment to reduce DOM elements.

File: blocks/product-gallery-block/features/product-gallery/default.jsx [hunk 8]
```diff
@@ -137,8 +153,6 @@ function ProductGallery({ customFields }) {
 	);
 }
 
-ProductGallery.label = "Product Gallery – Arc Block";
-
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/default.jsx]: Did you mean to remove this?

File: blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx (hunk 2 above)
Comment (ecarlisle) [reply blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx]: @nschubach : Good call! Made this change.

File: blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx (hunk 3 above)
Comment (ecarlisle) [reply blocks/product-gallery-block/features/product-gallery/_children/ProductFocusView/index.jsx]: @nschubach : Good call! Made this change.
I think the image `HY6LDPEW4BBFDLBYD4S3S7LZ3E` is coming from the original Product Gallery tests and that ID is being used multiple times, but I'm sure there's a way to make it unique.

File: blocks/product-gallery-block/features/product-gallery/default.jsx (hunk 8 above)
Comment (ecarlisle) [reply blocks/product-gallery-block/features/product-gallery/default.jsx]: Thanks! Unintended. I have replaced it.

File: blocks/product-gallery-block/features/product-gallery/default.jsx (hunk 7 above)
Comment (ecarlisle) [reply blocks/product-gallery-block/features/product-gallery/default.jsx]: Good call! Changed to a fragment.

**Comment by nschubach** [comment]
I know this is WIP, but are you adding a storybook item for this?

**Review by ecarlisle — COMMENTED** [review]
Refreshing comments if I started a review rather than adding comments.

**Comment by ecarlisle** [comment]
> I know this is WIP, but are you adding a storybook item for this?

@nschubach : Yes. I've listed that under the "Pending, Incomplete Items" section above. I increased the title size and placed it just before the Acceptance Criteria to give it more attention.

**Comment by JackHowa** [comment]
@ecarlisle 

I see the incomplete items. What have you seen not working here? what's expected? I'm also curious what you've tried so I can help you 

> 
> Pending, Incomplete Items
> The Product Focus View is not yet "Desktop Only." Click events on images need to render the overlay only when the viewport is desktop width or higher.
> Sass tokens have been set on default. Some/all may need to be moved to desktop depending on how the "Desktop Only" behavior is implemented.
> Storybook implementation of the Desktop view appears to be incomplete, though tested styles in arc-themes-feature-pack have also been placed in the Storybook commerce styles.
> The state of the thumbnail bar and selected thumbnail needs to change when a different product image scrolls is scrolled into view and "takes up the majority screen height." This is probably best accomplished using the Intersection Observer API.
> The number of Thumbnail images displayed depends on the height of the viewport. Currently that number is static and needs to be dynamic based on viewport height. is is probably also best accomplished using the Intersection Observer API.

**Comment by nschubach** [comment]
image
It seems that the collapse arrows on the Product-Content block are bleeding through... are we considering fixing that with this?

image

**Comment by nschubach** [comment]
Close button appears to have moved below the content?
image

**Comment by matthewroach** [comment]
@nschubach - Can you pull both the blocks branches and feature pack branches again - the issues you are seeing were addressed.

Feature Pack PR is - https://github.com/WPMedia/arc-themes-feature-pack/pull/359

File: blocks/product-gallery-block/features/product-gallery/_children/ThumbnailBar/index.jsx [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+	const thumbnailBarRef = useRef();
+	const shouldShowDownButton = () =>
+		images.length >= thumbnailBarItemLimit &&
+		thumbnailBarStartIndex + thumbnailBarItemLimit < images.length;
+
+	const shouldShowUpButton = () =>
+		thumbnailBarItemLimit < images.length && thumbnailBarStartIndex !== 0;
+
+	const windowResize = () => {
+		// Math
```
Comment (nschubach) [inline blocks/product-gallery-block/features/product-gallery/_children/ThumbnailBar/index.jsx]:
```suggestion
```

**Review by nschubach — APPROVED** [review]
Not sure I like the hardcoded desktop width (I think it could be just a styled `display:none` on this component) but this will do.
